### ISSUE TITLE
Convert popochiu 1.x project to version 2.0

### ISSUE THREAD

**Issue by anthonyirwin82** [carenalgas/popochiu#227]
This was discussed briefly in the discord chat and had a positive response, so I thought I would work on it.

I plan on doing this over two milestones because I believe that milestone one would be useful for end users and because it makes it easier for project maintainers to review the code.

# Milestone One (Convert Project Structure):
* Convert pascal case folders and filenames to snake case as used in popochiu 2.0
* Move files from res://popochiu to res://game
* Update the godot project file to have correct reference to the default scene.
* Update the res://game/popochiu_data.cfg file to have correct file references.
* Update the *.tres files to have correct file references.

# Milestone Two (Convert End User Code):
* Convert user written gdscript code to the Popochiu 2.0 version when version 1.x code has changed and is not backwards compatible.

I have already started to work on this and have significant progress for milestone one.

I have never used Popochiu 1.x so would ask that anyone who has worked at converting from 1.x to 2.0 post changes they had to make to their gdscript files in 2.0 in this post so that I know what changes need to be made for milestone two.

**Comment by mapedorr** [comment]
Sorry if I'm repeating myself but I can't stop feeling so touched and shocked about this: 🥳 **This is a terrific contribution for version 2 release!** 😃 I didn't see it coming.

Perhaps I won't be able to remember everything right now, but as I do, I'll add more comments.

### E.run is now E.queue

The calls to `E.run([` should be replaced by `E.queue([`. And in some cases they could be preceded by a `yield(`, so code like `yield(E.run([` should be changed to `await E.queue([`. But I'm not sure if Godot itself do that change (replacing calls to `yield` with the `await` keyword) when the project is converted.

Now, here comes a tricky part. Calls to some methods inside `E.run([` should be replaced by its `queue_` version. For example, taking the code [from this project](https://github.com/pointysalario/pointysalario), which is built in Popochiu 1.10:

```gdscript
E.run([
	C.walk_to_clicked(),
	C.face_clicked(),
	'Criptoevangelista: Bah, ese anuncio es truchisimo, no me vas a hacer caer en esa truchada.',
	'Criptoevangelista: Los verdaderos traders de MamṕṕónMusk tienen tarjetas oficiales de inversión',
])
```

In Popochiu 2, should be:

```gdscript
E.queue([
	C.queue_walk_to_clicked(),
	C.queue_face_clicked(),
	'Criptoevangelista: Bah, ese anuncio es truchisimo, no me vas a hacer caer en esa truchada.',
	'Criptoevangelista: Los verdaderos traders de MamṕṕónMusk tienen tarjetas oficiales de inversión',
])
```

Although it may ultimately be something that the developers themselves take care of.

### Replace on_interact and on_look

**PopochiuClickable** objects used to have the methods `on_interact` and `on_look`. Those are now called `_on_click` and `_on_right_click`.

### G.display is now G.show_system_text

Nothing to add about this.

### Remove the _now suffix

There were some methods with a `_now` suffix that should be removed. E.g. `add_now`, `add_as_active_now`, `remove_now`, `play_now`, `fade_now`, `stop_now`.

### Replace the _no_block suffix by _bg

Non blocking functions have now a `_bg` suffix instead of `_no_block`. E.g. `camera_shake_no_block` is now `camera_shake_bg`, `play_music_no_block` should be `play_music_bg`, 

> For this I'll have to add a couple of functions to the `i_audio.gd` and `i_character` scripts since I removed some methods from those because devs can access audio cues and character directly from code.

❗**We'll need to discuss on this point.**

### To take into account

Popochiu 1.x changed a lot during its development, specially since version 1.9, so it might happen that we should focus on making this to work on projects made with Popochiu 1.9 forwards, because updating older projects could be a gigantic work (and we don't know how many devs will be interested on that).

However, I believe that part of the script can be very useful for those who want to migrate a project they started in an alpha version.

### Extra comments

I saw in the video you shared on Discord that one of the reference projects is [popochius_little_adventure](https://github.com/mapedorr/popochius_little_adventure), that game was made in Popochiu 1.3.2, so it's likely too old to try migration.

Another project I saw in the video was [Zappin' da Mubis](https://github.com/mapedorr/advjam2023). In case you didn't know, it was made in Popochiu 2 (probably Alpha 1 or something earlier), so can be a good example of updating an Alpha project to Beta.

**Comment by stickgrinder** [comment]
@anthonyirwin82 I cannot but second @mapedorr! This contribution is terrific and will improve the DX of Popochiu **a lot!**

I would ask you if you can bake in a framework we can use to release automated upgrade paths in the future.

In my head, the Popochiu version should be stored somewhere in the project. If a user upgrades the engine in the addons folder, a set of procedures should be executed when the project is opened to bring it up to date with the changes.

This would allow us to add migration scripts with new versions as separate files and see them executed in order when someone does the upgrade.

I will need this for a feature I'm working on right now, and I planned to create a migration procedure by hand, but if you're working on it, we may as well find a common pattern.

Thanks a lot!

**Comment by anthonyirwin82** [comment]
I have created a migration framework that will automatically run all migrations one by one sequentially until all migrations are applied. The migrations are called from the res://addons/popochiu/popochiu_plugin.gd file without the user needing to do anything.

I added a res://addons/popochiu/migration folder that contains all migration specific things.

I added a [migration] section and a version key to the res://game/popochiu_data.cfg file so the migration framework knows what migrations need to be applied to the user project.

I have a migration config file that has the latest migration version. This means that the migrations are independent to the popochiu version. This is similar to how database migrations are specific to the database and not the software version. 

This way a popochiu migration can happen  during development and does not rely on the popochiu version as it looks at migration versions and if a migration exists it runs it.

**Comment by mapedorr** [comment]
That sounds pretty good, @anthonyirwin82. Yesterday, after reading @stickgrinder's comment, I was thinking that it would be nice if we could show a notification popup about the changes between the version of their project (I thought about storing that in the `popochiu_data.cfg` file) and the version of the plugin, and then they can trigger the tasks to migrate anything necessary from there.

**Comment by stickgrinder** [comment]
Hi @anthonyirwin82

Have you got a PR on the way for this issue? Do you think we can discuss having it released with Popochiu 2.0 stable in a couple of weeks?

**Comment by stickgrinder** [comment]
@mapedorr created a migration tool that is already equipped with everything necessary to upgrade from the first 2.0 alpha to the next stable.

We will use this tool from now on to ensure that versions upgrades will be automatic. We are not going to serve a 1.0 to 2.0 migration path because of the very few projects that would benefit from this.
It is not worth the effort.

Closing this issue for now.
